third_party/libunwindstack/MapInfo.cpp

```diff
@@ -176,6 +192,146 @@ Memory* MapInfo::GetFileMemory() {
+// for PE binaries with this property, we cannot simply associate an executable mapping to the
+// corresponding file using the path in the mapping.
+//
+// However, we can make an educated guess. The path of a PE will at least appear in the read-only
+// mapping that corresponds to the beginning of the file, which contains the headers (because the
+// offset in the file is zero and the address chosen for this mapping should always be a multiple of
+// the page size). Therefore, let's consider the file of the last file mapping that precedes this
+// anonymous mapping. If this file is a PE, and an executable file mapping for the .text section of
+// this PE is not present, and finally the offset and the size of this anonymous mapping are
+// compatible with the address range where the .text section would be loaded based on the header for
+// the section (in particular, VirtualAddress and VirtualSize), then we can be quite sure that this
+// anonymous mapping indeed contains the .text section of that PE. Note that we assume that a PE
+// only has one .text section and one executable mapping.
+//
+// This method is the counterpart of GetFileMemory for anonymous executable mappings and contains
+// the detection mechanism described.
+Memory* MapInfo::GetFileMemoryFromAnonExecMapIfPeCoffTextSection() {
+  if (!name().empty()) {
+    // The mapping already has a name.
+    return nullptr;
+  }
+  if ((flags() & PROT_EXEC) == 0) {
+    // The mapping is not executable.
+    return nullptr;
+  }
+
+  std::shared_ptr<MapInfo> map_info_it = prev_map();
+
+  // Note that when the first character of a map name is '[', the mapping is a special one like
+  // [stack], [heap], etc. Even if such mapping has a name, it's not file-backed. An alternative way
+  // of detecting that would be to check if the inode is 0 (but that's not available here).
+  while (map_info_it != nullptr &&
+         (map_info_it->name().empty() || map_info_it->name().c_str()[0] == '[')) {
+    map_info_it = map_info_it->prev_map();
+  }
+  if (map_info_it == nullptr) {
+    // There is no previous mapping with a name.
+    return nullptr;
+  }
+
+  if ((map_info_it->flags() & MAPS_FLAGS_DEVICE_MAP) != 0) {
+    // The previous file mapping corresponds to a character or block device.
+    return nullptr;
+  }
+
+  // This is the candidate name (file path).
+  std::string prev_name{map_info_it->name()};
+  CHECK(!prev_name.empty())
+
+  std::shared_ptr<MapInfo> first_map_info_with_prev_name = nullptr;
+
+  while (map_info_it != nullptr &&
+         (map_info_it->name().empty() || map_info_it->name().c_str()[0] == '[' ||
+          map_info_it->name() == prev_name)) {
+    if (map_info_it->name() == prev_name) {
+      if ((map_info_it->flags() & PROT_EXEC) != 0) {
+        // There is already an executable mapping for this name.
+        return nullptr;
+      }
+      first_map_info_with_prev_name = map_info_it;
```

Comment:
So why again do we expect multiple file named mappings with the same name in our case? Because the alignment difference could only occur later?

Reply:
It could happen that the .text section is at an offset that doesn't allow a file mapping but one preceding section is, in addition to the headers.

Reply:
Could you add a brief comment?

Reply:
Yeah, this can happen if the file alignment for a section coincidentally is also page aligned. And there could be multiple of those potentially. I agree a  comment would be good.

Reply:
Yep, done (before the loop).